File: ruby_test/ruby-worker/ping_worker.rb
require 'net/http'
require 'uri'
require 'json'
require 'webrick'
require 'thread'

# Service model
class Service
  attr_accessor :id, :name, :url, :interval
  
  def initialize(id, name, url, interval)
    @id = id
    @name = name
    @url = url
    @interval = interval
  end
end

# The Ping Worker
class PingWorker
  def initialize(services)
    @services = services
  end

  def ping(service)
    start_time = Time.now
    uri = URI.parse(service.url)
    
    begin
      response = Net::HTTP.get_response(uri)
      latency = (Time.now - start_time) * 1000 # convert to ms
      
      if response.is_a?(Net::HTTPSuccess)
        puts "[#{Time.now.strftime('%H:%M:%S')}] #{service.id}: online (#{latency.round(2)}ms)"
      else
        puts "[#{Time.now.strftime('%H:%M:%S')}] #{service.id}: offline (HTTP #{response.code})"
      end
    rescue => e
      puts "[#{Time.now.strftime('%H:%M:%S')}] #{service.id}: offline (Error: #{e.message})"
    end
  end

  def run
    puts "🚀 PulsePing Ruby Worker started. Monitoring #{@services.length} services..."
    loop do
      threads = @services.map do |service|
        Thread.new { ping(service) }
      end
      
      threads.each(&:join)
      puts "--------------------------------------------------"
      sleep 15 # Check every 15 seconds for this demo
    end
  end
end

# Sample Data
services = [
  Service.new("api-gateway", "API Gateway", "https://api.github.com", 300),
  Service.new("web-frontend", "Web Frontend", "https://google.com", 900)
]

worker = PingWorker.new(services)

# Start Monitoring in a Background Thread
Thread.new { worker.run }

# Start a simple WEBrick server for Render Health Checks
port = ENV['PORT'] || 8080
server = WEBrick::HTTPServer.new(Port: port, AccessLog: [], Logger: WEBrick::Log.new(File::NULL))

server.mount_proc '/' do |req, res|
  res.body = "PulsePing Ruby Worker is active!"
end

server.mount_proc '/health' do |req, res|
  res.status = 200
  res.body = "OK"
end

trap('INT') { server.shutdown }

puts "🌍 Web server started on port #{port}"
server.start

File: java_test/src/main/java/com/pulseping/PingWorker.java
package com.pulseping;

import com.sun.net.httpserver.HttpServer;
import java.io.IOException;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

public class PingWorker {

    static class Service {
        String id;
        String name;
        String url;

        Service(String id, String name, String url) {
            this.id = id;
            this.name = name;
            this.url = url;
        }
    }

    private static final HttpClient client = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(10))
            .build();

    private static final DateTimeFormatter dtf = DateTimeFormatter.ofPattern("HH:mm:ss");

    public static void main(String[] args) throws IOException {
        List<Service> services = List.of(
                new Service("api-gateway", "API Gateway", "https://api.github.com"),
                new Service("web-frontend", "Web Frontend", "https://google.com")
        );

        // Start Monitoring Thread
        new Thread(() -> runMonitoringLoop(services)).start();

        // Start Health Check Server
        int port = Integer.parseInt(System.getenv().getOrDefault("PORT", "8080"));
        HttpServer server = HttpServer.create(new InetSocketAddress(port), 0);
        
        server.createContext("/", (t) -> {
            String response = "PulsePing Java Worker is active!";
            t.sendResponseHeaders(200, response.length());
            try (OutputStream os = t.getResponseBody()) {
                os.write(response.getBytes());
            }
        });

        server.createContext("/health", (t) -> {
            String response = "OK";
            t.sendResponseHeaders(200, response.length());
            try (OutputStream os = t.getResponseBody()) {
                os.write(response.getBytes());
            }
        });

        System.out.println("🌍 Web server started on port " + port);
        server.start();
    }

    private static void runMonitoringLoop(List<Service> services) {
        System.out.println("🚀 PulsePing Java Worker started. Monitoring " + services.size() + " services...");
        ExecutorService executor = Executors.newFixedThreadPool(services.size());

        while (true) {
            for (Service service : services) {
                executor.submit(() -> ping(service));
            }

            try {
                System.out.println("--------------------------------------------------");
                TimeUnit.SECONDS.sleep(15);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                break;
            }
        }
    }

    private static void ping(Service service) {
        long start = System.currentTimeMillis();
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(service.url))
                .header("User-Agent", "PulsePing-Monitor/1.0")
                .GET()
                .build();

        try {
            HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
            long latency = System.currentTimeMillis() - start;

            if (response.statusCode() < 400) {
                System.out.printf("[%s] %s: online (%dms)%n", dtf.format(LocalDateTime.now()), service.id, latency);
            } else {
                System.out.printf("[%s] %s: offline (HTTP %d)%n", dtf.format(LocalDateTime.now()), service.id, response.statusCode());
            }
        } catch (Exception e) {
            System.out.printf("[%s] %s: offline (Error: %s)%n", dtf.format(LocalDateTime.now()), service.id, e.getMessage());
        }
    }
}
